pkg/core/dns/cache.go

```diff
@@ -0,0 +1,39 @@
+package dns
+
+import (
+	"net"
+	"sync"
+	"time"
+)
+
+type LookupIPFunc func(string) ([]net.IP, error)
+
+type cacheRecord struct {
+	ips          []net.IP
+	creationTime time.Time
+}
+
+func MakeCaching(f LookupIPFunc, ttl time.Duration) LookupIPFunc {
```

Comment:
I think we can figure out better name. Maybe `CachedLookupIP`? `CachedLookupIPFunc`?